File: src/m/classes/bamgmesh.py
import numpy as np


class bamgmesh(object):
    """BAMGMESH class definition

    Usage:
        bamgmesh(varargin)
    """

    def __init__(self, *args):  # {{{
        self.Vertices = np.empty((0, 3))
        self.Edges = np.empty((0, 3))
        self.Triangles = np.empty((0, 0))
        self.IssmEdges = np.empty((0, 0))
        self.IssmSegments = np.empty((0, 0))
        self.VerticesOnGeomVertex = np.empty((0, 0))
        self.VerticesOnGeomEdge = np.empty((0, 0))
        self.EdgesOnGeomEdge = np.empty((0, 0))
        self.SubDomains = np.empty((0, 4))
        self.SubDomainsFromGeom = np.empty((0, 0))
        self.ElementConnectivity = np.empty((0, 0))
        self.NodalConnectivity = np.empty((0, 0))
        self.NodalElementConnectivity = np.empty((0, 0))
        self.CrackedVertices = np.empty((0, 0))
        self.CrackedEdges = np.empty((0, 0))

        if not len(args):
            # if no input arguments, create a default object
            pass

        elif len(args) == 1:
            object = args[0]
            for field in list(object.keys()):
                if field in vars(self):
                    setattr(self, field, object[field])

        else:
            raise TypeError("bamgmesh constructor error message: unknown type of constructor call")
    # }}}

    def __repr__(self):  # {{{
        s = "class '%s' object '%s' = \n" % (type(self), 'self')
        s += "    Vertices: %s\n" % str(self.Vertices)
        s += "    Edges: %s\n" % str(self.Edges)
        s += "    Triangles: %s\n" % str(self.Triangles)
        s += "    IssmEdges: %s\n" % str(self.IssmEdges)
        s += "    IssmSegments: %s\n" % str(self.IssmSegments)
        s += "    VerticesOnGeomVertex: %s\n" % str(self.VerticesOnGeomVertex)
        s += "    VerticesOnGeomEdge: %s\n" % str(self.VerticesOnGeomEdge)
        s += "    EdgesOnGeomEdge: %s\n" % str(self.EdgesOnGeomEdge)
        s += "    SubDomains: %s\n" % str(self.SubDomains)
        s += "    SubDomainsFromGeom: %s\n" % str(self.SubDomainsFromGeom)
        s += "    ElementConnectivity: %s\n" % str(self.ElementConnectivity)
        s += "    NodalConnectivity: %s\n" % str(self.NodalConnectivity)
        s += "    NodalElementConnectivity: %s\n" % str(self.NodalElementConnectivity)
        s += "    CrackedVertices: %s\n" % str(self.CrackedVertices)
        s += "    CrackedEdges: %s\n" % str(self.CrackedEdges)
        return s
    # }}}

File: src/c/bamg/BamgMesh.cpp
#include "./bamgobjects.h"
#include "../shared/shared.h"

/*Constructors/Destructors*/
BamgMesh::BamgMesh(){/*{{{*/

	this->VerticesSize[0]=0,                  this->VerticesSize[1]=0;                 this->Vertices=NULL;          this->PreviousNumbering = NULL;
	this->EdgesSize[0]=0,                     this->EdgesSize[1]=0;                    this->Edges=NULL;
	this->TrianglesSize[0]=0,                 this->TrianglesSize[1]=0;                this->Triangles=NULL;

	this->SubDomainsSize[0]=0,                this->SubDomainsSize[1]=0;               this->SubDomains=NULL;
	this->SubDomainsFromGeomSize[0]=0,        this->SubDomainsFromGeomSize[1]=0;       this->SubDomainsFromGeom=NULL;
	this->CrackedVerticesSize[0]=0,           this->CrackedVerticesSize[1]=0;          this->CrackedVertices=NULL;
	this->CrackedEdgesSize[0]=0,              this->CrackedEdgesSize[1]=0;             this->CrackedEdges=NULL;

	this->VerticesOnGeomVertexSize[0]=0,      this->VerticesOnGeomVertexSize[1]=0;     this->VerticesOnGeomVertex=NULL;
	this->VerticesOnGeomEdgeSize[0]=0,        this->VerticesOnGeomEdgeSize[1]=0;       this->VerticesOnGeomEdge=NULL;
	this->EdgesOnGeomEdgeSize[0]=0,           this->EdgesOnGeomEdgeSize[1]=0;          this->EdgesOnGeomEdge=NULL;

	this->IssmEdgesSize[0]=0,                 this->IssmEdgesSize[1]=0;                this->IssmEdges=NULL;
	this->IssmSegmentsSize[0]=0,              this->IssmSegmentsSize[1]=0;             this->IssmSegments=NULL;

	this->ElementConnectivitySize[0]=0,       this->ElementConnectivitySize[1]=0;      this->ElementConnectivity=NULL;
	this->NodalConnectivitySize[0]=0,         this->NodalConnectivitySize[1]=0;        this->NodalConnectivity=NULL;
	this->NodalElementConnectivitySize[0]=0,  this->NodalElementConnectivitySize[1]=0; this->NodalElementConnectivity=NULL;
}
/*}}}*/
BamgMesh::~BamgMesh(){/*{{{*/

	xDelete<double>(this->Vertices);
	xDelete<double>(this->PreviousNumbering);
	xDelete<double>(this->Edges);
	xDelete<double>(this->Triangles);

	xDelete<double>(this->SubDomains);
	xDelete<double>(this->SubDomainsFromGeom);
	xDelete<double>(this->CrackedVertices);
	xDelete<double>(this->CrackedEdges);

	xDelete<double>(this->VerticesOnGeomVertex);
	xDelete<double>(this->VerticesOnGeomEdge);
	xDelete<double>(this->EdgesOnGeomEdge);

	xDelete<double>(this->IssmEdges);
	xDelete<double>(this->IssmSegments);

	xDelete<double>(this->ElementConnectivity);
	xDelete<double>(this->NodalConnectivity);
	xDelete<double>(this->NodalElementConnectivity);
}
/*}}}*/
